Sources/SwiftParser/Lexer.swift

```diff
@@ -310,6 +303,91 @@ extension Lexer {
   }
 }
 
+// MARK: - Peeking
+
+extension Lexer.Cursor {
+  func peek(at offset: Int = 0) -> UInt8? {
+    assert(offset >= 0)
+    guard offset < self.input.count else {
```

Comment:
It would be good to check the performance cost of running this on every single peek, ie. do we still need an unsafe version for when we know we're not at EOF?

Reply:
Performance measurements of `testSelfParsePerformance` (release builds):

- Current `main`: 6.061 s
- With https://github.com/apple/swift-syntax/pull/1192: 6.285 (+3.7%)
- This PR: 6.330 (+ 0.7%)

I think this is acceptable given the added safety this gives us.

<details>
<summary>Raw data</summary>

```
main

Test Case '-[SwiftParserTest.ParserTests testSelfParsePerformance]' measured [Time, seconds] average: 6.062, relative standard deviation: 1.030%, values: [6.188462, 6.127593, 6.119602, 6.032848, 6.075716, 6.015036, 5.986559, 6.063117, 6.017368, 5.995290], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , polarity: prefers smaller, maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[SwiftParserTest.ParserTests testSelfParsePerformance]' measured [Time, seconds] average: 6.061, relative standard deviation: 0.619%, values: [6.136058, 6.019346, 6.036355, 6.052679, 6.034555, 6.046183, 6.101317, 6.109356, 6.043739, 6.031751], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , polarity: prefers smaller, maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100


lex-string-delimiters

Test Case '-[SwiftParserTest.ParserTests testSelfParsePerformance]' measured [Time, seconds] average: 6.298, relative standard deviation: 0.360%, values: [6.314035, 6.311459, 6.294189, 6.284121, 6.263693, 6.275958, 6.329505, 6.316271, 6.323488, 6.268774], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , polarity: prefers smaller, maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
Test Case '-[SwiftParserTest.ParserTests testSelfParsePerformance]' measured [Time, seconds] average: 6.272, relative standard deviation: 0.664%, values: [6.369508, 6.263280, 6.288378, 6.267686, 6.229722, 6.224330, 6.228591, 6.259549, 6.308582, 6.278303], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , polarity: prefers smaller, maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100


This PR

Test Case '-[SwiftParserTest.ParserTests testSelfParsePerformance]' measured [Time, seconds] average: 6.340, relative standard deviation: 0.672%, values: [6.456374, 6.327926, 6.361954, 6.321446, 6.342133, 6.335581, 6.328410, 6.287928, 6.318708, 6.322288], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage
Test Case '-[SwiftParserTest.ParserTests testSelfParsePerformance]' measured [Time, seconds] average: 6.319, relative standard deviation: 0.740%, values: [6.443643, 6.314616, 6.300575, 6.312995, 6.347714, 6.276697, 6.307492, 6.329889, 6.280654, 6.279295], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , polarity: prefers smaller, maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100
```

</details>

Reply:
I'm somewhat surprised by the ~4% fro the delimiters split. Any idea why that's so high?

Reply:
I just measured again using Instruments (and a command line tool that contains the code from `testSelfParsePerformance` instead of `testSelfParsePerformance` itself) and couldn’t find a significant difference between any of the branches. Not sure what introduced the variation before.